src/tests/roc_lib/test_sender_receiver.cpp

```diff
@@ -116,24 +117,41 @@ class Sender : public core::Thread {
 
 private:
     virtual void run() {
-        for (size_t off = 0; off < total_samples_; off += frame_size_) {
-            if (off + frame_size_ > total_samples_) {
-                off = total_samples_ - frame_size_;
+        float signal_sample = sample_step_;
+        float samples[total_samples_];
+
+        // TODO need a condition ?
+        while (true) {
+            for (size_t i = 0; i < total_samples_; ++i) {
+                samples[i] = signal_sample;
             }
 
-            roc_frame frame;
-            memset(&frame, 0, sizeof(frame));
+            roc_log(LogInfo, "Original sample #%lu: %f", (unsigned long)0,
+                    (double)samples[0]);
 
-            frame.samples = samples_ + off;
-            frame.samples_size = frame_size_ * sizeof(float);
+            for (size_t off = 0; off < total_samples_; off += frame_size_) {
+                if (off + frame_size_ > total_samples_) {
+                    off = total_samples_ - frame_size_;
+                }
+
+                roc_frame frame;
+                memset(&frame, 0, sizeof(frame));
+
+                frame.samples = samples + off;
+                frame.samples_size = frame_size_ * sizeof(float);
+
+                const int ret = roc_sender_write(sndr_, &frame);
+                roc_panic_if_not(ret == 0);
+            }
 
-            const int ret = roc_sender_write(sndr_, &frame);
-            roc_panic_if_not(ret == 0);
+            // TODO Bad idea to sleep ?
+            core::sleep_until(core::timestamp() + 100 * core::Millisecond);
+            signal_sample += sample_step_;
```

Comment:
Yep, it's a bad idea :)

Sender and receiver are real-time entities. Sender produces samples according to the specified sample rate (44100) and receiver expects that samples will arrive according to the rate as well (in average).

Some info is available here: https://roc-project.github.io/roc/docs/internals/data_flow.html#pipeline-timing

There are two modes how sender or receiver can function:

* *Clocked by user*. If `config.automatic_timing` is 0, the user is responsible to pass the samples according to the specified sample rate. This option is useful when the user reads samples from or writes sample to an audio device which has its own clock. In this case the sender or receiver will be clocked by this device as well.

* *Clocked by a CPU-timer*. If `config.automatic_timing` is 1, the write (for sender) or read (for receiver) operation will automatically block until it's time to send the next bunch of samples according to a CPU-timer. This option is useful when the user reads samples from or writes sample to a file which does not have its own clock. In this case the sender or receiver will be clocked by its own internal timer.

Why do we need this option? Because the second option (a CPU timer) can't be used with audio devices. An audio device and a CPU timer, even both configured to, say, 41100 Hz, will belong to different clock domains (CPU and soundcard) and will have slightly different clocks. One of them will be a bit faster or slower (in a few hertz).

So we should allow the user to disable the CPU timer to prevent underruns/overruns. On the other hand, when the user works with audio files, not devices, it would be convenient to do the clocking automatically for the user. Thus we have this option. See more details here: https://roc-project.github.io/roc/docs/internals/fe_resampler.html

In these tests `automatic_timing` is set to 1. It means that the sender will automatically block us when necessary. We should not perform any sleeping by ourselves. Just generate more samples and block on roc_sender_write() until it decides that it's time to send them.

Why do we need this sleep here?

Reply:
Because at first i thought that the sample rate was respected **during** each call of `roc_sender_write` but not **between** the calls. So I wanted to add a sleep at the end of the while loop to make Sender respects the sample rate before calling again `roc_sender_write`.

Reply:
Ah, I see. Nope, the CPU timer state is preserved between the write calls. Probably we need to make the documentation more clear on this, I'll add add this to my TODO.